src/IntrusivePtr.h

```diff
@@ -135,7 +135,7 @@ struct IntrusivePtr {
     }
 
     IntrusivePtr<T> &operator=(const IntrusivePtr<T> &other) {
-        if (other.ptr == ptr) {
+        if (this == &other) {
```

Comment:
This is different though. The old code compared by value, and the new code compares by identity. So in the old code, more cases could do the early-out.

Reply:
...yes, this is true, but in practice, is this something that actually happens frequently enough to worry about?

Reply:
It happens whenever you assign an Expr to another Expr that happens to already be the same thing. E.g.:

```
Expr my_function(Expr a, Expr b) {
 ...
 a = b;
 ...
}
...
my_function(x, x);
```

This includes uninitialized Exprs:
```
Thing make_a_thing(... const Expr &e ... ) {
  Thing thing; // has default-initialized Expr fields
  ...
  thing.foo = e;
  return thing;
}
...
make_a_thing(..., Expr(), ...)
```

I inserted a check for `other.ptr == ptr && this != &other` and compiled local laplacian. It triggered 51016 times. 43k were the non-null case and 7k were the null case.

Reply:
...that's worth a //NOLINT.

Reply:
Probably also worth a link to this discussion to reduce future confusion.

Reply:
Done